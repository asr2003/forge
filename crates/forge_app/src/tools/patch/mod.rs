mod apply;
mod apply_json;
mod marker;
mod parse;

pub use apply_json::ApplyPatchJson;
