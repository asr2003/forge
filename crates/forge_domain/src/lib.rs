mod agent;
mod chat_request;
mod chat_response;
mod context;
mod conversation;
mod env;
mod error;
mod event;
mod file;
mod message;
mod model;
mod orch;
mod point;
mod provider;
mod suggestion;
mod summarize;
mod template;
mod tool;
mod tool_call;
mod tool_call_parser;
mod tool_choice;
mod tool_definition;
mod tool_name;
mod tool_result;
mod tool_usage;
mod workflow;

pub use agent::*;
pub use chat_request::*;
pub use chat_response::*;
pub use context::*;
pub use conversation::*;
pub use env::*;
pub use error::*;
pub use event::*;
pub use file::*;
pub use message::*;
pub use model::*;
pub use orch::*;
pub use point::*;
pub use provider::*;
pub use suggestion::*;
pub use summarize::*;
pub use template::*;
pub use tool::*;
pub use tool_call::*;
pub use tool_call_parser::*;
pub use tool_choice::*;
pub use tool_definition::*;
pub use tool_name::*;
pub use tool_result::*;
pub use tool_usage::*;
pub use workflow::*;

#[async_trait::async_trait]
pub trait ProviderService: Send + Sync + 'static {
    async fn chat(
        &self,
        id: &ModelId,
        context: Context,
    ) -> ResultStream<ChatCompletionMessage, anyhow::Error>;
    async fn models(&self) -> anyhow::Result<Vec<Model>>;
    async fn parameters(&self, model: &ModelId) -> anyhow::Result<Parameters>;
}

#[async_trait::async_trait]
pub trait ToolService: Send + Sync {
    // TODO: should take `call` by reference
    async fn call(&self, call: ToolCallFull) -> ToolResult;
    fn list(&self) -> Vec<ToolDefinition>;
    fn usage_prompt(&self) -> String;
}

#[async_trait::async_trait]
pub trait ConversationService: Send + Sync {
    async fn get(&self, id: &ConversationId) -> anyhow::Result<Option<Conversation>>;
    async fn create(&self, workflow: Workflow) -> anyhow::Result<ConversationId>;
    async fn inc_turn(&self, id: &ConversationId, agent: &AgentId) -> anyhow::Result<()>;
    async fn set_context(
        &self,
        id: &ConversationId,
        agent: &AgentId,
        context: Context,
    ) -> anyhow::Result<()>;
    async fn insert_event(
        &self,
        conversation_id: &ConversationId,
        event: Event,
    ) -> anyhow::Result<()>;
}

#[async_trait::async_trait]
pub trait TemplateService: Send + Sync {
    async fn render_system(
        &self,
        agent: &Agent,
        prompt: &Template<SystemContext>,
    ) -> anyhow::Result<String>;

    async fn render_event(
        &self,
        agent: &Agent,
        prompt: &Template<EventContext>,
        event: &Event,
    ) -> anyhow::Result<String>;
}

/// Core app trait providing access to services and repositories.
/// This trait follows clean architecture principles for dependency management
/// and service/repository composition.
pub trait App: Send + Sync + 'static {
    type ToolService: ToolService;
    type ProviderService: ProviderService;
    type ConversationService: ConversationService;
    type TemplateService: TemplateService;

    fn tool_service(&self) -> &Self::ToolService;
    fn provider_service(&self) -> &Self::ProviderService;
    fn conversation_service(&self) -> &Self::ConversationService;
    fn template_service(&self) -> &Self::TemplateService;
}
