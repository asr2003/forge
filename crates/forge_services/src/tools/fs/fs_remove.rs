use std::path::Path;
use std::sync::Arc;

use forge_domain::{
    ExecutableTool, FSRemoveInput, NamedTool, ToolCallContext, ToolDescription, ToolName,
};
use forge_tool_macros::ToolDescription;

use crate::tools::utils::assert_absolute_path;
use crate::{FileRemoveService, FsMetaService, Infrastructure};

// Using FSRemoveInput from forge_domain

/// Request to remove a file at the specified path. Use this when you need to
/// delete an existing file. The path must be absolute. This operation cannot
/// be undone, so use it carefully.
#[derive(ToolDescription)]
pub struct FSRemove<T>(Arc<T>);

impl<T: Infrastructure> FSRemove<T> {
    pub fn new(infra: Arc<T>) -> Self {
        Self(infra)
    }
}

impl<T> NamedTool for FSRemove<T> {
    fn tool_name() -> ToolName {
        ToolName::new("forge_tool_fs_remove")
    }
}

#[async_trait::async_trait]
impl<T: Infrastructure> ExecutableTool for FSRemove<T> {
    type Input = FSRemoveInput;

    async fn call(&self, _context: ToolCallContext, input: Self::Input) -> anyhow::Result<String> {
        let path = Path::new(&input.path);
        assert_absolute_path(path)?;

        // Check if the file exists
        if !self.0.file_meta_service().exists(path).await? {
            return Err(anyhow::anyhow!("File not found: {}", input.path));
        }

        // Check if it's a file
        if !self.0.file_meta_service().is_file(path).await? {
            return Err(anyhow::anyhow!("Path is not a file: {}", input.path));
        }

        // Remove the file
        self.0.file_remove_service().remove(path).await?;

        Ok(format!("Successfully removed file: {}", input.path))
    }
}

#[cfg(test)]
mod test {
    use bytes::Bytes;

    use super::*;
    use crate::attachment::tests::MockInfrastructure;
    use crate::tools::utils::TempDir;
    use crate::{FsCreateDirsService, FsWriteService};

    #[tokio::test]
    async fn test_fs_remove_success() {
        let temp_dir = TempDir::new().unwrap();
        let file_path = temp_dir.path().join("test.txt");
        let infra = Arc::new(MockInfrastructure::new());

        // Create a test file
        infra
            .file_write_service()
            .write(
                file_path.as_path(),
                Bytes::from("test content".as_bytes().to_vec()),
            )
            .await
            .unwrap();

        assert!(infra.file_meta_service().exists(&file_path).await.unwrap());

        let fs_remove = FSRemove::new(infra.clone());
        let result = fs_remove
            .call(
                ToolCallContext::default(),
                FSRemoveInput { path: file_path.to_string_lossy().to_string() },
            )
            .await
            .unwrap();

        assert!(result.contains("Successfully removed file"));
        assert!(!infra.file_meta_service().exists(&file_path).await.unwrap());
    }

    #[tokio::test]
    async fn test_fs_remove_nonexistent_file() {
        let temp_dir = TempDir::new().unwrap();
        let nonexistent_file = temp_dir.path().join("nonexistent.txt");
        let infra = Arc::new(MockInfrastructure::new());

        let fs_remove = FSRemove::new(infra);
        let result = fs_remove
            .call(
                ToolCallContext::default(),
                FSRemoveInput { path: nonexistent_file.to_string_lossy().to_string() },
            )
            .await;

        assert!(result.is_err());
        assert!(result.unwrap_err().to_string().contains("File not found"));
    }

    #[tokio::test]
    async fn test_fs_remove_directory() {
        let temp_dir = TempDir::new().unwrap();
        let dir_path = temp_dir.path().join("test_dir");
        let infra = Arc::new(MockInfrastructure::new());

        // Create a test directory
        infra
            .create_dirs_service()
            .create_dirs(dir_path.as_path())
            .await
            .unwrap();
        assert!(infra
            .file_meta_service()
            .exists(dir_path.as_path())
            .await
            .unwrap());

        let fs_remove = FSRemove::new(infra.clone());
        let result = fs_remove
            .call(
                ToolCallContext::default(),
                FSRemoveInput { path: dir_path.to_string_lossy().to_string() },
            )
            .await;

        assert!(result.is_err());
        assert!(result
            .unwrap_err()
            .to_string()
            .contains("Path is not a file"));
        assert!(infra
            .file_meta_service()
            .exists(dir_path.as_path())
            .await
            .unwrap());
    }

    #[tokio::test]
    async fn test_fs_remove_relative_path() {
        let infra = Arc::new(MockInfrastructure::new());
        let fs_remove = FSRemove::new(infra);
        let result = fs_remove
            .call(
                ToolCallContext::default(),
                FSRemoveInput { path: "relative/path.txt".to_string() },
            )
            .await;

        assert!(result.is_err());
        assert!(result
            .unwrap_err()
            .to_string()
            .contains("Path must be absolute"));
    }
}
