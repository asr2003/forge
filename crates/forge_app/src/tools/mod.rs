mod fetch;
mod fs;
mod patch;
mod shell;
mod syn;
mod think;
mod utils;

use std::sync::Arc;

use fetch::Fetch;
use forge_domain::Tool;
use fs::*;
use patch::*;
use shell::Shell;
use think::Think;

use crate::{EnvironmentService, Infrastructure};

pub fn tools<F: Infrastructure>(infra: Arc<F>) -> Vec<Tool> {
    let env = infra.environment_service().get_environment();
    vec![
        FSRead.into(),
        FSWrite.into(),
        FSRemove.into(),
        FSList::default().into(),
        FSSearch.into(),
        FSFileInfo.into(),
        // TODO: once ApplyPatchJson is stable we can delete ApplyPatch
        // ApplyPatch.into(),
        ApplyPatchJson.into(),
        Shell::new(env.clone()).into(),
        Think::default().into(),
        Fetch::default().into(),
    ]
}

#[cfg(test)]
mod tests {
    use std::path::{Path, PathBuf};

    use forge_domain::{Environment, Point, Query, Suggestion};

    use super::*;
    use crate::{EmbeddingService, FileReadService, VectorIndex};

    /// Create a default test environment
    fn stub() -> Stub {
        Stub {
            env: Environment {
                os: std::env::consts::OS.to_string(),
                cwd: std::env::current_dir().unwrap_or_default(),
                home: Some("/".into()),
                shell: if cfg!(windows) {
                    "cmd.exe".to_string()
                } else {
                    "/bin/sh".to_string()
                },
                base_path: PathBuf::new(),
                qdrant_key: Default::default(),
                qdrant_cluster: Default::default(),
                pid: std::process::id(),
                provider_url: Default::default(),
                provider_key: Default::default(),
                openai_key: Default::default(),
            },
        }
    }

    struct Stub {
        env: Environment,
    }

    #[async_trait::async_trait]
    impl EmbeddingService for Stub {
        async fn embed(&self, _text: &str) -> anyhow::Result<Vec<f32>> {
            unimplemented!()
        }
    }

    #[async_trait::async_trait]
    impl EnvironmentService for Stub {
        fn get_environment(&self) -> Environment {
            self.env.clone()
        }
    }
    #[async_trait::async_trait]
    impl FileReadService for Stub {
        async fn read(&self, _path: &Path) -> anyhow::Result<String> {
            unimplemented!()
        }
    }
    #[async_trait::async_trait]
    impl VectorIndex<Suggestion> for Stub {
        async fn store(&self, _information: Point<Suggestion>) -> anyhow::Result<()> {
            unimplemented!()
        }

        async fn search(&self, _query: Query) -> anyhow::Result<Vec<Point<Suggestion>>> {
            unimplemented!()
        }
    }

    #[async_trait::async_trait]
    impl Infrastructure for Stub {
        type EnvironmentService = Stub;
        type FileReadService = Stub;
        type VectorIndex = Stub;
        type EmbeddingService = Stub;

        fn environment_service(&self) -> &Self::EnvironmentService {
            self
        }

        fn file_read_service(&self) -> &Self::FileReadService {
            self
        }

        fn vector_index(&self) -> &Self::VectorIndex {
            self
        }

        fn embedding_service(&self) -> &Self::EmbeddingService {
            self
        }
    }

    #[test]
    fn test_tool_description_length() {
        const MAX_DESCRIPTION_LENGTH: usize = 1024;

        println!("\nTool description lengths:");

        let mut any_exceeded = false;
        let stub = Arc::new(stub());
        for tool in tools(stub.clone()) {
            let desc_len = tool.definition.description.len();
            println!(
                "{:?}: {} chars {}",
                tool.definition.name,
                desc_len,
                if desc_len > MAX_DESCRIPTION_LENGTH {
                    "(!)"
                } else {
                    ""
                }
            );

            if desc_len > MAX_DESCRIPTION_LENGTH {
                any_exceeded = true;
            }
        }

        assert!(
            !any_exceeded,
            "One or more tools exceed the maximum description length of {}",
            MAX_DESCRIPTION_LENGTH
        );
    }
}
