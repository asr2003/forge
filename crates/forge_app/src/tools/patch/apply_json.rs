use std::path::Path;

// No longer using dissimilar for fuzzy matching
use forge_domain::{ExecutableTool, NamedTool, ToolDescription, ToolName};
use forge_tool_macros::ToolDescription;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use thiserror::Error;
use tokio::fs;

use crate::tools::syn;
use crate::tools::utils::assert_absolute_path;

// Removed fuzzy matching threshold as we only use exact matching now

/// A match found in the source text. Represents a range in the source text that
/// can be used for extraction or replacement operations. Stores the position
/// and length to allow efficient substring operations.
#[derive(Debug, Clone, Copy, Eq, PartialEq, Ord, PartialOrd)]
struct Range {
    /// Starting position of the match in source text
    start: usize,
    /// Length of the matched text
    length: usize,
}

impl Range {
    /// Create a new match from a start position and length
    fn new(start: usize, length: usize) -> Self {
        Self { start, length }
    }

    /// Get the end position (exclusive) of this match
    fn end(&self) -> usize {
        self.start + self.length
    }

    /// Try to find an exact match in the source text
    fn find_exact(source: &str, search: &str) -> Option<Self> {
        source
            .find(search)
            .map(|start| Self::new(start, search.len()))
    }

    // Fuzzy matching removed - we only use exact matching
}

impl From<Range> for std::ops::Range<usize> {
    fn from(m: Range) -> Self {
        m.start..m.end()
    }
}

// MatchSequence struct and implementation removed - we only use exact matching

#[derive(Debug, Error)]
enum Error {
    #[error("Failed to read/write file: {0}")]
    FileOperation(#[from] std::io::Error),
    #[error("Could not find match for search text: {0}")]
    NoMatch(String),
    #[error("Could not find swap target text: {0}")]
    NoSwapTarget(String),
}

fn apply_replacement(
    source: String,
    search: &str,
    operation: &Operation,
    content: &str,
) -> Result<String, Error> {
    // Handle empty search string - only certain operations make sense here
    if search.is_empty() {
        return match operation {
            // Append to the end of the file
            Operation::Append => Ok(format!("{}{}", source, content)),
            // Prepend to the beginning of the file
            Operation::Prepend => Ok(format!("{}{}", content, source)),
            // Replace is equivalent to completely replacing the file
            Operation::Replace => Ok(content.to_string()),
            // Swap doesn't make sense with empty search - keep source unchanged
            Operation::Swap => Ok(source),
        };
    }

    // Find the exact match to operate on
    let patch =
        Range::find_exact(&source, search).ok_or_else(|| Error::NoMatch(search.to_string()))?;

    // Apply the operation based on its type
    match operation {
        // Prepend content before the matched text
        Operation::Prepend => Ok(format!(
            "{}{}{}",
            &source[..patch.start],
            content,
            &source[patch.start..]
        )),

        // Append content after the matched text
        Operation::Append => Ok(format!(
            "{}{}{}",
            &source[..patch.end()],
            content,
            &source[patch.end()..]
        )),

        // Replace matched text with new content
        Operation::Replace => Ok(format!(
            "{}{}{}",
            &source[..patch.start],
            content,
            &source[patch.end()..]
        )),

        // Swap with another text in the source
        Operation::Swap => {
            // Find the target text to swap with
            let target_patch = Range::find_exact(&source, content)
                .ok_or_else(|| Error::NoSwapTarget(content.to_string()))?;

            // Handle the case where patches overlap
            if (patch.start <= target_patch.start && patch.end() > target_patch.start)
                || (target_patch.start <= patch.start && target_patch.end() > patch.start)
            {
                // For overlapping ranges, we just do an ordinary replacement
                return Ok(format!(
                    "{}{}{}",
                    &source[..patch.start],
                    content,
                    &source[patch.end()..]
                ));
            }

            // We need to handle different ordering of patches
            if patch.start < target_patch.start {
                // Original text comes first
                Ok(format!(
                    "{}{}{}{}{}",
                    &source[..patch.start],
                    content,
                    &source[patch.end()..target_patch.start],
                    &source[patch.start..patch.end()],
                    &source[target_patch.end()..]
                ))
            } else {
                // Target text comes first
                Ok(format!(
                    "{}{}{}{}{}",
                    &source[..target_patch.start],
                    &source[patch.start..patch.end()],
                    &source[target_patch.end()..patch.start],
                    content,
                    &source[patch.end()..]
                ))
            }
        }
    }
}

/// Operation types that can be performed on matched text
#[derive(Deserialize, Serialize, JsonSchema, Debug, Clone, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum Operation {
    /// Prepend content before the matched text
    Prepend,

    /// Append content after the matched text
    Append,

    /// Replace the matched text with new content
    Replace,

    /// Swap the matched text with another text (search for the second text and
    /// swap them)
    Swap,
}

#[derive(Deserialize, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub struct ApplyPatchJsonInput {
    /// The path to the file to modify
    pub path: String,

    /// The text to search for in the source. If empty, operation applies to the
    /// end of the file.
    pub search: String,

    /// The operation to perform on the matched text
    pub operation: Operation,

    /// The content to use for the operation (replacement text, text to
    /// prepend/append, or target text for swap operations)
    pub content: String,
}

/// Performs a single text operation (prepend, append, replace, swap, delete) on
/// matched text in a file. The operation is applied to the first match found in
/// the text.
#[derive(ToolDescription)]
pub struct ApplyPatchJson;

impl NamedTool for ApplyPatchJson {
    fn tool_name() -> ToolName {
        ToolName::new("tool_forge_fs_patch")
    }
}

/// Format the modified content as XML with optional syntax warning
fn format_output(path: &str, content: &str, warning: Option<&str>) -> String {
    if let Some(w) = warning {
        format!(
            "<file_content\n  path=\"{}\"\n  syntax_checker_warning=\"{}\">\n{}</file_content>\n",
            path, w, content
        )
    } else {
        format!(
            "<file_content path=\"{}\">\n{}\n</file_content>\n",
            path,
            content.trim_end()
        )
    }
}

/// Process the file modification and return the formatted output
async fn process_file_modifications(
    path: &Path,
    search: &str,
    operation: &Operation,
    content: &str,
) -> Result<String, Error> {
    let file_content = fs::read_to_string(path).await?;
    let file_content = apply_replacement(file_content, search, operation, content)?;
    fs::write(path, &file_content).await?;

    let warning = syn::validate(path, &file_content).map(|e| e.to_string());
    Ok(format_output(
        path.to_string_lossy().as_ref(),
        &file_content,
        warning.as_deref(),
    ))
}

#[async_trait::async_trait]
impl ExecutableTool for ApplyPatchJson {
    type Input = ApplyPatchJsonInput;

    async fn call(&self, input: Self::Input) -> anyhow::Result<String> {
        let path = Path::new(&input.path);
        assert_absolute_path(path)?;

        Ok(
            process_file_modifications(path, &input.search, &input.operation, &input.content)
                .await?,
        )
    }
}

#[cfg(test)]
mod test {

    use super::*;

    // Enhanced test helper for running multiple operations
    #[derive(Debug)]
    struct PatchTest {
        initial: String,
        patches: Vec<Patch>,
    }

    // Single operation with its result
    #[derive(Debug)]
    struct Patch {
        operation: PatchOperation,
        result: Result<String, String>,
    }

    // Represents a single patch operation
    #[derive(Debug)]
    struct PatchOperation {
        search: String,
        operation: Operation,
        content: String,
    }

    // fmt::Display implementation removed in favor of using assert_debug_snapshot!

    impl PatchTest {
        fn new(initial: impl ToString) -> Self {
            PatchTest { initial: initial.to_string(), patches: Vec::new() }
        }

        /// Replace matched text with new content
        fn replace(mut self, search: impl ToString, content: impl ToString) -> Self {
            let operation = PatchOperation {
                search: search.to_string(),
                operation: Operation::Replace,
                content: content.to_string(),
            };
            self.patches.push(Patch {
                operation,
                result: Err("Not executed yet".to_string()), // Placeholder
            });
            self
        }

        /// Prepend content before matched text
        fn prepend(mut self, search: impl ToString, content: impl ToString) -> Self {
            let operation = PatchOperation {
                search: search.to_string(),
                operation: Operation::Prepend,
                content: content.to_string(),
            };
            self.patches.push(Patch {
                operation,
                result: Err("Not executed yet".to_string()), // Placeholder
            });
            self
        }

        /// Append content after matched text
        fn append(mut self, search: impl ToString, content: impl ToString) -> Self {
            let operation = PatchOperation {
                search: search.to_string(),
                operation: Operation::Append,
                content: content.to_string(),
            };
            self.patches.push(Patch {
                operation,
                result: Err("Not executed yet".to_string()), // Placeholder
            });
            self
        }

        /// Swap matched text with target text
        fn swap(mut self, search: impl ToString, target: impl ToString) -> Self {
            let operation = PatchOperation {
                search: search.to_string(),
                operation: Operation::Swap,
                content: target.to_string(),
            };
            self.patches.push(Patch {
                operation,
                result: Err("Not executed yet".to_string()), // Placeholder
            });
            self
        }

        /// Try to execute all operations and record their results
        fn execute_all(mut self) -> Self {
            let mut current_content = self.initial.clone();

            for op_result in &mut self.patches {
                // Apply the operation
                let result = match apply_replacement(
                    current_content.clone(),
                    &op_result.operation.search,
                    &op_result.operation.operation,
                    &op_result.operation.content,
                ) {
                    Ok(content) => {
                        // Update the current content for the next operation
                        current_content = content.clone();
                        Ok(content)
                    }
                    Err(err) => Err(err.to_string()),
                };

                // Update the result
                op_result.result = result;
            }

            self
        }
    }

    #[test]
    fn comprehensive_patch_tests() {
        // Create a comprehensive test that includes all the test cases
        let test = PatchTest::new("Hello World")
            // Basic Operations
            .replace("World", "Forge")
            .replace("", " bar")
            // Single Replacement Behavior
            .replace("foo", "baz")
            // Exact Matching Behavior
            .replace("Hello", "Hi")
            // Unicode and Special Characters
            .replace("Hello", "你好")
            .replace("World", "🌍")
            // Whitespace Handling
            .prepend("Hello", "    ")
            .append("World", "\n  New line")
            // Test different operation types
            .prepend("Hello", "Greetings, ")
            .append("World", "!")
            .swap("Hello", "World")
            // Empty search operations
            .prepend("", "Start: ")
            .append("", " End")
            .replace("", "Completely New Content")
            // Execute all operations and collect results
            .execute_all();

        // Snapshot the entire test result using Debug representation
        insta::assert_debug_snapshot!(test);
    }

    #[test]
    fn comprehensive_error_tests() {
        // Create a test specifically for error cases
        let test = PatchTest::new("foo bar baz")
            .replace("nonexistent", "replaced")
            .replace("foo-bar", "replaced")
            .replace("afoo", "replaced")
            .swap("foo", "nonexistent")
            .execute_all();

        // Snapshot the error test results using Debug representation
        insta::assert_debug_snapshot!(test);
    }

    // The previous individual tests are removed since they're now consolidated
}
