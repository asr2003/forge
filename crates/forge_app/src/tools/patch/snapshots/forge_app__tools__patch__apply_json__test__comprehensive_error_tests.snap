---
source: crates/forge_app/src/tools/patch/apply_json.rs
expression: test
snapshot_kind: text
---
PatchTest {
    initial: "foo bar baz",
    patches: [
        Patch {
            operation: PatchOperation {
                search: "nonexistent",
                operation: Replace,
                content: "replaced",
            },
            result: Err(
                "Could not find match for search text: nonexistent",
            ),
        },
        Patch {
            operation: PatchOperation {
                search: "foo-bar",
                operation: Replace,
                content: "replaced",
            },
            result: Err(
                "Could not find match for search text: foo-bar",
            ),
        },
        Patch {
            operation: PatchOperation {
                search: "afoo",
                operation: Replace,
                content: "replaced",
            },
            result: Err(
                "Could not find match for search text: afoo",
            ),
        },
        Patch {
            operation: PatchOperation {
                search: "foo",
                operation: Swap,
                content: "nonexistent",
            },
            result: Err(
                "Could not find swap target text: nonexistent",
            ),
        },
    ],
}
